File: app/src/main/java/cn/ftoutiao/account/android/activity/login/LoginPresenter.java
package cn.ftoutiao.account.android.activity.login;

import android.app.Activity;
import android.support.annotation.NonNull;

import com.acmenxd.logger.Logger;
import com.acmenxd.retrofit.exception.HttpException;
import com.acmenxd.toaster.Toaster;
import com.umeng.socialize.UMAuthListener;
import com.umeng.socialize.UMShareAPI;
import com.umeng.socialize.bean.SHARE_MEDIA;

import java.util.Map;

import cn.ftoutiao.account.android.base.BasePresenter;
import cn.ftoutiao.account.android.constants.ConstanPool;
import cn.ftoutiao.account.android.constants.UrlConstans;
import cn.ftoutiao.account.android.model.NoteBookEntity;
import cn.ftoutiao.account.android.model.SaltEntity;
import cn.ftoutiao.account.android.model.UserEntity;
import cn.ftoutiao.account.android.utils.MD5Utils;

/**
 * Created by weichyang on 2018/2/9.
 */

public class LoginPresenter extends BasePresenter<LoginContract.View> implements LoginContract.Presenter, UMAuthListener {


    /**
     * 构造器,传入BaseView实例
     *
     * @param pView
     */
    public LoginPresenter(LoginContract.View pView) {
        super(pView);
    }

    @Override
    public void login(final String name, final String pwd) {
        request().initlogin(UrlConstans.BASE_URL + UrlConstans.REQUEST_SALT, name, ConstanPool.TYPE_LOGIN)
                .enqueue(new BindCallback<SaltEntity>() {

                    @Override
                    public void succeed(@NonNull SaltEntity pData) {
                        super.succeed(pData);
                        doLogin(name, pwd, pData);
                    }

                    @Override
                    public void failed(@NonNull HttpException pE) {
                        super.failed(pE);
                        Toaster.show(pE.getMsg());
                    }
                });


    }

    /**
     * 登录
     *
     * @param name
     * @param pwd
     * @param saltEntity
     */
    public void doLogin(String name, String pwd, SaltEntity saltEntity) {
        mView.showLoadingDialog(true);
        request().dologin(UrlConstans.BASE_URL + UrlConstans.REQUEST_LOGIN, name, MD5Utils.getPwdHash(pwd, saltEntity.salt))
                .enqueue(new BindCallback<UserEntity>() {
                    @Override
                    public void succeed(@NonNull UserEntity pData) {
                        super.succeed(pData);
                        mView.hideLoadingDialog();
                        mView.loginSuccess(pData);
                    }

                    @Override
                    public void failed(@NonNull HttpException pE) {
                        super.failed(pE);
                        mView.hideLoadingDialog();
                        mView.loginFailed(pE.getMsg());
                    }
                });


    }

    @Override
    public void oauthLogin(SHARE_MEDIA media, Activity activity) {
        UMShareAPI.get(activity).getPlatformInfo(activity, media, this);
    }

    @Override
    public void requestNoteBookCategory(String aid) {
        request().requestNotebookCatagory(
                UrlConstans.BASE_URL + UrlConstans.REQUEST_BOOKCATEGORY,
                aid).enqueue(
                new BindCallback<NoteBookEntity>() {
                    @Override
                    public void succeed(@NonNull NoteBookEntity pData) {
                        super.succeed(pData);
                        mView.requestNoteBookCategorySuccess(pData);
                    }

                    @Override
                    public void failed(@NonNull HttpException pE) {
                        super.failed(pE);
                        Toaster.show(pE.getMsg());
                        mView.requestFaliur();
                    }
                });
    }


// 三方登录

    @Override
    public void onStart(SHARE_MEDIA share_media) {
        Logger.d("--start");
    }

    @Override
    public void onComplete(SHARE_MEDIA share_media, int i, Map<String, String> map) {
        if (share_media != null) {
            String id = "";
            String accessToken = "";
            String platform = "";
            String unionId = "";
            String refreshToken = "";

            switch (share_media) {
                case SINA:
                    id = map.get("uid");
                    accessToken = map.get("access_token");
                    platform = "sina";
                    refreshToken = map.get("refresh_token");
                    break;
                case QQ:
                    id = map.get("openid");
                    accessToken = map.get("access_token");
                    platform = "qq";
                    break;
                case WEIXIN:
                    id = map.get("openid");
                    unionId = map.get("unionid");
                    accessToken = map.get("access_token");
                    platform = "weixin";
                    refreshToken = map.get("refreshToken");
                    break;
                default:
            }

            request().authorLogin(UrlConstans.BASE_URL + UrlConstans.REQUEST_AUTHLOGIN,
                    id, accessToken, platform, refreshToken, unionId)
                    .enqueue(new BindCallback<UserEntity>() {
                        @Override
                        public void succeed(@NonNull UserEntity pData) {
                            super.succeed(pData);
                            mView.loginSuccess(pData);
                        }

                        @Override
                        public void failed(@NonNull HttpException pE) {
                            Toaster.show(pE.getMessage());
                        }
                    });

        }
    }

    @Override
    public void onError(SHARE_MEDIA share_media, int i, Throwable throwable) {
        Logger.d("--onerroe");
        mView.loginFailed(throwable.getMessage());
    }

    @Override
    public void onCancel(SHARE_MEDIA share_media, int i) {
        Logger.d("--onCancel");
    }
}

File: app/src/main/java/cn/ftoutiao/account/android/activity/login/LoginPresenter.kt
//package cn.ftoutiao.account.android.login
//
//import android.app.Activity
//import cn.ftoutiao.account.android.core.entity.Account
//import cn.ftoutiao.account.android.core.entity.Salt
//import cn.ftoutiao.account.android.core.retrofit.BookResult
//import cn.ftoutiao.account.android.login.datasource.AccountDataSource
//import com.umeng.socialize.UMAuthListener
//import com.umeng.socialize.UMShareAPI
//import com.umeng.socialize.bean.SHARE_MEDIA
//import io.reactivex.android.schedulers.AndroidSchedulers
//import io.reactivex.schedulers.Schedulers
//import retrofit2.Call
//import retrofit2.Callback
//import retrofit2.Response
//
//
///**
// * Created by alan on 2017/11/7.
// */
//class LoginPresenter(private var mView: LoginContract.View, private var mDataSource: AccountDataSource)
//    : LoginContract.Presenter, UMAuthListener {
//
//
//
//    override fun login(name: String, pwd: String) {
//        mView.showLoading()
//        mDataSource.getSalt(name, Salt.TYPE_LOGIN)
//                .subscribeOn(Schedulers.io())
//                .flatMap { saltResult -> mDataSource.login(name, pwd, saltResult.salt) }
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe({ accountResult ->
//                    run {
//                        mView.hideLoading()
//                        mView.loginSuccess(accountResult)
//                    }
//                }) { t ->
//                    run {
//                        mView.hideLoading()
//                        mView.loginFailed(t.message!!)
//                        t.printStackTrace()
//                    }
//                }
//    }
//
//    override fun oauthLogin(media: SHARE_MEDIA, activity: Activity) {
//        UMShareAPI.get(activity).getPlatformInfo(activity, media, this)
//    }
//
//    override fun onStart(p0: SHARE_MEDIA?) {
//
//    }
//
//    override fun onComplete(p0: SHARE_MEDIA?, p1: Int, p2: MutableMap<String, String>?) {
//
//        if (p0 != null) {
//            var id: String? = null
//            var accessToken: String? = null
//            var platform = ""
//            var unionId: String? = null
//            var refreshToken: String? = null
//            when (p0) {
//                SHARE_MEDIA.SINA -> {
//                    id = p2?.get("uid")
//                    accessToken = p2?.get("access_token")
//                    platform = "sina"
//                    refreshToken = p2?.get("refresh_token")
//                }
//
//                SHARE_MEDIA.QQ -> {
//                    id = p2?.get("openid")
//                    accessToken = p2?.get("access_token")
//                    platform = "qq"
//                }
//
//                SHARE_MEDIA.WEIXIN -> {
//                    id = p2?.get("openid")
//                    unionId = p2?.get("unionid")
//                    accessToken = p2?.get("access_token")
//                    platform = "weixin"
//                    refreshToken = p2?.get("refreshToken")
//                }
//                else -> {
//                }
//            }
//            mDataSource.oauthLogin(id!!, accessToken!!, platform, unionId, refreshToken)
//                    .enqueue(object : Callback<BookResult<Account>> {
//                        override fun onResponse(call: Call<BookResult<Account>>?, response: Response<BookResult<Account>>?) {
//                            mView.loginSuccess(response?.body()!!)
//                        }
//
//                        override fun onFailure(call: Call<BookResult<Account>>?, t: Throwable?) {
//                            mView.loginFailed(t?.message!!)
//                        }
//                    })
//        }
//    }
//
//    override fun onCancel(p0: SHARE_MEDIA?, p1: Int) {
//
//    }
//
//    override fun onError(p0: SHARE_MEDIA?, p1: Int, p2: Throwable?) {
//        mView.loginFailed(p2?.message!!)
//    }
//}